src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs

```diff
@@ -329,6 +330,108 @@ public bool IsNodeAddedRecently
+        /// Gets a value indicating whether this model is frozen.
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this instance is frozen; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsFrozen
+        {
+            get
+            {              
+                return NodeModel.IsFrozen;
+            }
+            set
+            {
+                NodeModel.IsFrozen = value;
+            }
+        }
+        
+        /// <summary>
+        /// Gets a value indicating whether Freeze property on the node is checked
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if node is not frozen and CanExecute; otherwise, <c>false</c>.
+        /// </value>        
+        public bool NodeRunChecked
+        {
+            get
+            {    
+                //if the node is freeze by the user, then always
+                //check the Freeze property     
+                if (this.NodeLogic.explictFrozen)
+                {                   
+                    return true;
+                }
+                
+                return false;
+            }             
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the freeze property is enabled.
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if [node run enabled]; otherwise, <c>false</c>.
+        /// </value>
+        public bool NodeRunEnabled
+        {
+            get
+            {
+                //this is the default case.
+                if (!this.NodeLogic.explictFrozen &&
+                      !this.NodeLogic.IsFrozen)
+                {
+                    return true;
+                }
+
+                //If any of the node is set to freeze by the user and 
+                // if that node is frozen by itself, then disable the Freeze property
+                var checklist = new List<NodeModel>();
+                CheckIfAnyParentAndChildAreExplictlyFrozen(this.NodeLogic,checklist);
+                if (this.nodeLogic.explictFrozen && checklist.Any())
+                {
+                    return false;
+                }
+                               
+                //if the node is set to freeze by the user     
+                // then enable the Freeze property
+                if (this.NodeLogic.explictFrozen)                   
+                {
+                    return true;
+                }
+                                
+                return false;
+            }
+            
+        }
+
+        /// <summary>
+        /// For the given node, this traverses the graph upstream to check
+        /// if any of the node in upstream is explictly frozen.
+        /// </summary>
+        /// <param name="node">The node.</param>
+        /// <param name="checkNodesList">The check nodes list.</param>
+        private void CheckIfAnyParentAndChildAreExplictlyFrozen(NodeModel node,List<NodeModel> checkNodesList )
```

Comment:
Is this a duplication of the method that you added to `NodeModel` to do the graph analysis?